go/oasis-test-runner/scenario/e2e/runtime/txsource.go

```diff
@@ -485,10 +505,35 @@ func (sc *txSourceImpl) Run(childEnv *env.Env) error {
 	}
 
 	// Start all configured workloads.
-	errCh := make(chan error, len(sc.workloads)+2)
-	for _, name := range sc.workloads {
-		if err := sc.startWorkload(childEnv, errCh, name); err != nil {
-			return fmt.Errorf("failed to start workload %s: %w", name, err)
+	errCh := make(chan error, len(sc.clientWorkloads)+len(sc.allNodeWorkloads)+2)
+	for _, name := range sc.clientWorkloads {
+		if err := sc.startWorkload(childEnv, errCh, name, &sc.Net.Clients()[0].Node); err != nil {
+			return fmt.Errorf("failed to start client workload %s: %w", name, err)
+		}
+	}
+	for _, name := range sc.allNodeWorkloads {
+		// Determine nodes that we should run the workload on.
+		var nodes []*oasis.Node
+		for _, v := range sc.Net.Validators() {
```

Comment:
Maybe this can be extracted into a method of the `Network` struct, in case it'll be useful elsewhere.
Also consider moving this out of the above loop.

Reply:
Good idea, will refactor this a bit.